scripts/exregional_process_bufr.sh

```diff
@@ -0,0 +1,386 @@
+scrfunc_fn=$( basename "${scrfunc_fp}" )
+scrfunc_dir=$( dirname "${scrfunc_fp}" )
+#
+#-----------------------------------------------------------------------
+#
+# Print message indicating entry into script.
+#
+#-----------------------------------------------------------------------
+#
+print_info_msg "
+========================================================================
+Entering script:  \"${scrfunc_fn}\"
+In directory:     \"${scrfunc_dir}\"
+
+This is the ex-script for the task that runs bufr (cloud, metar, lightning) preprocess
+with FV3 for the specified cycle.
+========================================================================"
+#
+#-----------------------------------------------------------------------
+#
+# Extract from CDATE the starting year, month, day, and hour of the
+# forecast.  These are needed below for various operations.
+#
+#-----------------------------------------------------------------------
+#
+START_DATE=$(echo "${CDATE}" | sed 's/\([[:digit:]]\{2\}\)$/ \1/')
+YYYYMMDDHH=$(date +%Y%m%d%H -d "${START_DATE}")
+JJJ=$(date +%j -d "${START_DATE}")
+
+YYYY=${YYYYMMDDHH:0:4}
+MM=${YYYYMMDDHH:4:2}
+DD=${YYYYMMDDHH:6:2}
+HH=${YYYYMMDDHH:8:2}
+YYYYMMDD=${YYYYMMDDHH:0:8}
+
+YYJJJHH=$(date +"%y%j%H" -d "${START_DATE}")
+PREYYJJJHH=$(date +"%y%j%H" -d "${START_DATE} 1 hours ago")
+
+#
+#-----------------------------------------------------------------------
+#
+# Get into working directory
+#
+#-----------------------------------------------------------------------
+#
+print_info_msg "$VERBOSE" "
+Getting into working directory for BUFR observation process ..."
+
+cd_vrfy ${DATA}
+
+fixgriddir=$FIXgsi/${PREDEF_GRID_NAME}
+
+print_info_msg "$VERBOSE" "fixgriddir is $fixgriddir"
+
+#
+#-----------------------------------------------------------------------
+#
+# link or copy background files
+#
+#-----------------------------------------------------------------------
+
+cp_vrfy ${fixgriddir}/fv3_grid_spec          fv3sar_grid_spec.nc
+
+#-----------------------------------------------------------------------
+#
+#   copy bufr table
+#
+#-----------------------------------------------------------------------
+BUFR_TABLE=${FIXgsi}/prepobs_prep_RAP.bufrtable
+cp_vrfy $BUFR_TABLE prepobs_prep.bufrtable
+
+#-----------------------------------------------------------------------
+#
+#   set observation soruce 
+#
+#-----------------------------------------------------------------------
+if [[ "${NET}" = "RTMA"* ]]; then
+  SUBH=$(date +%M -d "${START_DATE}")
+  obs_source="rtma_ru"
+  obsfileprefix=${obs_source}
+  obspath_tmp=${OBSPATH}/${obs_source}.${YYYYMMDD}
+
+else
+  SUBH=""
+  obs_source=rap
+  if [[ ${HH} -eq '00' || ${HH} -eq '12' ]]; then
+    obs_source=rap_e
+  fi
+
+  case $MACHINE in
+
+  "WCOSS2")
+
+    obsfileprefix=${obs_source}
+    obspath_tmp=${OBSPATH}/${obs_source}.${YYYYMMDD}
+
+    ;;
+  *)
+
+    obsfileprefix=${YYYYMMDDHH}.${obs_source}
+    obspath_tmp=${OBSPATH}
+
```

Comment:
I think this type of information should live in the appropriate machine file, moving us away from any machine case statements.

Reply:
I am afraid this is inevitable. For example it would be awkward  to put the below code in a machine file. Even if it is possible, I don't think it is a good idea to put such a large piece of code in a machine file.

https://github.com/ufs-community/ufs-srweather-app/blob/2fd2c059388e9d6d33911ca5b1ced14e757703dc/scripts/exregional_run_anal.sh#L197-L267

Reply:
I disagree that case statements are "inevitable". The example you point out here is largely identical code with just a few changes; it might require some refactoring but I can't imagine this couldn't be handled without a `case` statement. 

In the case of the current PR, it would be as simple as making `obsfileprefix` a variable read in from the machine file.

Reply:
@mkavulich Note that for this PR, `obsfileprefix` is a function of a variable defined in the script itself namely `obs_source` for non-wcoss2 machines. That depends on the cycle hour HH. I am not saying it is not possible to do it, but is it worth it? The way I see it we will end up with some cryptic code embedded in a string, far from where it is used, so it will be. It may be acceptable for something short and generic (used in most ex-scripts) like `PRE_TASK_CMDS` that mostly sets `ulimit -a`, but I don't think it is a good idea to do this routinely for everything machine specific.

Reply:
I'm seeing now, too, that `OBSPATH` hasn't been defined in this PR. 

We don't yet have a task in the workflow to get the bufr observations for input to this task, but we do have some of the bufr observations locations defined in machine files already. Ensuring that the bufr files are retrieved appropriately via that avenue should remove the need for any of this logic in this script. The only thing this script will need is the location that SRW stages the files in the experiment directory once retrieve_data.py is run to fetch them.

Reply:
I have templated the path to be `OBSFILE_TEMPLATE` that contains the path and prefix. The default value covers most machines, for WCOSS2 specifically, it is defined in the corresponding machine file.
I have tested the task again in the bigger PR by incorporating these new changes.

Reply:
OBSPATH is defined in the high-priority PR #647 that this PR depends on. FIXgsi too, however I have mentioned these dependencies in the j-job description.